### PR TITLE
gitbase/repository_pool: change repository to be an interface

### PR DESCRIPTION
**Pull request by jfontan**
Also add `billyRepo` for billy filesystems.

I'm not totally happy with the naming of `repository` interface as there's already a `Repository` structure and all the `gitRepository`/`gitRepo` for struct and constructor. Suggestions welcome.

### CODE REVIEW

**Comment by erizocosmico** [comment]
Why do we need to add billy repositories to gitbase? IMHO, testing some more filesystem errors might not be worth introducing all this exclusively for testing.

**Comment by jfontan** [comment]
@erizocosmico moved `billyRepository` to the test file (`fs_error_test.go`)

File: fs_error_test.go [hunk 1]
```diff
@@ -81,6 +81,25 @@ func setupErrorRepos(t *testing.T) (*sql.Context, CleanupFunc) {
...
+	brokenType brokenType,
+	fs billy.Filesystem,
+) (billy.Filesystem, error) {
+	dotFS, err := fs.Chroot(".git")
+	if err != nil {
+		return nil, err
+	}
+
+	var brokenFS billy.Filesystem
+	if brokenType == 0 {
```
Comment (kuba--) [inline fs_error_test.go]: what's _0_? do we need a const for this?

**Comment by erizocosmico** [comment]
What I mean is: do we really benefit from making Repository an interface and the complexity this comes with just to test some FS errors?

**Comment by jfontan** [comment]
Even if we don't add more types of repositories and move to other ways to test errors (maybe fixtures with broken packfiles) I still think it makes sense. It puts the different logic for siva and git repos in one place instead of different code paths in other parts:

https://github.com/src-d/gitbase/pull/384/files#diff-b7e4ef15e429a4c21afc4a8822b635beL129

File: fs_error_test.go (hunk 1 above)
Comment (jfontan) [reply fs_error_test.go]: It means no broken flags, I've just added a constant.

**Comment by jfontan** [comment]
@ajnavarro I've just rebased

**Pull request merged**